src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs

```diff
@@ -113,8 +114,15 @@ private static bool MatchesLookupInfo(TagHelperDescriptor descriptor, LookupInfo
                 return false;
             }
 
-            return string.IsNullOrEmpty(lookupInfo.TypeName) ||
-              string.Equals(descriptor.TypeName, lookupInfo.TypeName, StringComparison.Ordinal);
+            // We need to escape the TypePattern so we can choose to only allow specific regex.
+            var escaped = Regex.Escape(lookupInfo.TypePattern);
+
+            // We surround the escaped with ^ and $ in order ot ensure a regex match matches the entire 
+            // string. We also replace any '*' or '?' characters with regex to match appropriate content.
+            var pattern = "^" + escaped.Replace(@"\?", ".?").Replace(@"\*", ".*?") + "$";
```

Comment:
`".*?` so we don't over-match what users are looking for (lazy evaluation). As for your second comment as to `'?'` during the code review I thought it was optional, which after now doing some more research realize it is not (too used to regex meaning): https://github.com/aspnet/Razor/issues/297

As for Razor author discovery, that's a problem with all Razor directives currently and is why this issue exists: https://github.com/aspnet/Razor/issues/291